sys-utils/libcmt/include/libcmt/keccak.h

```diff
@@ -0,0 +1,147 @@
+/* Copyright Cartesi and individual authors (see AUTHORS)
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @file
+ * @defgroup libcmt_keccak keccak
+ * Keccak 256 digest
+ *
+ * Data can be inserted in pieces:
+ * @code
+ * ...
+ * uint8_t h[CMT_KECCAK_LENGTH];
+ * cmt_keccak_t st[1];
+ *
+ * cmt_keccak_init(st);
+ * cmt_keccak_update(st, 1, "h");
+ * cmt_keccak_update(st, 1, "e");
+ * cmt_keccak_update(st, 1, "l");
+ * cmt_keccak_update(st, 1, "l");
+ * cmt_keccak_update(st, 1, "o");
+ * cmt_keccak_final(st, h);
+ * ...
+ * @endcode
+ *
+ * all at once:
+ * @code
+ * ...
+ * const char data[] = "hello";
+ * uint8_t h[CMT_KECCAK_LENGTH];
+ * cmt_keccak_data(h, sizeof(data)-1, data);
+ * ...
+ * @endcode
+ *
+ * or with a specialized call to generate @ref funsel data:
+ * @code
+ * ...
+ * uint32_t funsel = cmt_keccak_funsel("FunctionCall(address,bytes)");
+ * ...
+ * @endcode
+ *
+ * Code based on https://github.com/mjosaarinen/tiny_sha3 with the sha3 to
+ * keccak change as indicated by this comment:
+ * https://github.com/mjosaarinen/tiny_sha3#updated-07-aug-15
+ *
+ * @ingroup libcmtcmt
+ * @{ */
+#ifndef CMT_KECCAK_H
+#define CMT_KECCAK_H
+#include<stdint.h>
+#include<stddef.h>
+
+/** Opaque Keccak state, used to do hash computations, initialize with:
+ * - @ref cmt_keccak_init
+ * - @ref CMT_KECCAK_INIT
+ * - @ref CMT_KECCAK_DECL */
+typedef struct {
+    union {
```

Comment:
This union could get a typedef name, and its declaration moved outside. Unnamed struct/unions in the public C API are not friendly with C binding generators.